src/execution/compiler/operator/index_create_translator.cpp

```diff
@@ -0,0 +1,228 @@
+      table_pr_(codegen_->MakeFreshIdentifier("table_pr")),
+      tvi_var_(codegen_->MakeFreshIdentifier("tvi")),
+      vpi_var_(codegen_->MakeFreshIdentifier("vpi")),
+      col_oids_var_(codegen_->MakeFreshIdentifier("col_oids")),
+      slot_var_(codegen_->MakeFreshIdentifier("slot")),
+      table_schema_(codegen_->GetCatalogAccessor()->GetSchema(GetPlanAs<planner::CreateIndexPlanNode>().GetTableOid())),
+      all_oids_(AllColOids(table_schema_)),
+      table_pm_(codegen_->GetCatalogAccessor()
+                    ->GetTable(GetPlanAs<planner::CreateIndexPlanNode>().GetTableOid())
+                    ->ProjectionMapForOids(all_oids_)),
+      index_oid_() {
+  pipeline->RegisterSource(this, Pipeline::Parallelism::Serial);
+}
+
+void IndexCreateTranslator::PerformPipelineWork(WorkContext *context, FunctionBuilder *function) const {
+  InitScan(function);
+
+  PrepareContext(context, function);
+  // Scan it.
+  ScanTable(context, function);
+
+  // Close TVI, if need be.
+  function->Append(codegen_->TableIterClose(codegen_->MakeExpr(tvi_var_)));
+
+  FreeInserter(function);
+}
+
+void IndexCreateTranslator::InitScan(FunctionBuilder *function) const {
+  // Init inserter, index pr, table pr, create index, and init tvi
+  DeclareInserter(function);
+  CreateIndex(function);
+  DeclareIndexPR(function);
+  DeclareTablePR(function);
+  DeclareTVI(function);
+  DeclareSlot(function);
+}
+
+void IndexCreateTranslator::DeclareInserter(FunctionBuilder *function) const {
+  // var col_oids: [num_cols]uint32
+  // col_oids[i] = ...
+  SetOids(function);
+  // var inserter : StorageInterface
+  auto *storage_interface_type = codegen_->BuiltinType(ast::BuiltinType::Kind::StorageInterface);
+  function->Append(codegen_->DeclareVar(inserter_, storage_interface_type, nullptr));
+  // @storageInterfaceInit(inserter, execCtx, table_oid, col_oids_var_, true)
+  ast::Expr *inserter_setup = codegen_->StorageInterfaceInit(
+      inserter_, GetExecutionContext(), uint32_t(GetPlanAs<planner::CreateIndexPlanNode>().GetTableOid()),
+      col_oids_var_, false);
+  function->Append(codegen_->MakeStmt(inserter_setup));
+}
+
+void IndexCreateTranslator::SetOids(FunctionBuilder *function) const {
+  // var col_oids_var_: [num_cols]uint32
+  ast::Expr *arr_type = codegen_->ArrayType(all_oids_.size(), ast::BuiltinType::Kind::Uint32);
+  function->Append(codegen_->DeclareVar(col_oids_var_, arr_type, nullptr));
+
+  for (uint16_t i = 0; i < all_oids_.size(); i++) {
+    // col_oids_var_[i] = col_oid
+    ast::Expr *lhs = codegen_->ArrayAccess(col_oids_var_, i);
+    ast::Expr *rhs = codegen_->Const32(uint32_t(all_oids_[i]));
+    function->Append(codegen_->Assign(lhs, rhs));
+  }
+}
+
+void IndexCreateTranslator::CreateIndex(FunctionBuilder *function) const {
+  auto plan_node = common::ManagedPointer<const planner::CreateIndexPlanNode>(
+      dynamic_cast<const planner::CreateIndexPlanNode *>(Op()));
+  auto result = sql::DDLExecutors::CreateIndexExecutor(
```

Comment:
Do we want to create the index while generating TPL code? It seems more correct to have a TPL builtin, @createIndex(...) which returns a boolean. You would then use the boolean here.

Reply:
Now index is created before generating TPL code